### PR TITLE
[WIP] Use Messages

### PR DESCRIPTION
**Pull request by PythooonUser**
Not entirely done yet. Uses `LookAtDTO` to transfer data of the currently looked at object to the UI.

### CODE REVIEW

File: Dungeoneer/src/com/interrupt/dungeoneer/dto/LookAtDTO.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package com.interrupt.dungeoneer.dto;
+
+import com.badlogic.gdx.graphics.Color;
+
+public class LookAtDTO {
```
Comment (joshuaskelly) [inline Dungeoneer/src/com/interrupt/dungeoneer/dto/LookAtDTO.java:5]: Should rename this to `LookAtInfo` 
1. I like that name more (I had to look up what a DTO was)
2. All the methods to get the object call it a LookAtInfo 😄

File: Dungeoneer/src/com/interrupt/dungeoneer/entities/Player.java [hunk 2]
```diff
@@ -240,6 +238,9 @@
     // Used to act on breaking changes between save versions
     public int saveVersion = -1;
 
+    /** Currently focused object. */
+    public LookAt lookedAtItem = null;
```
Comment (joshuaskelly) [inline Dungeoneer/src/com/interrupt/dungeoneer/entities/Player.java:242]: Probably can shorten this to just `lookedAt` or `lookedAtEntity`? Because it could be more than just an `Item` object.

File: Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/GameMessage.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            }
+        };
+
+        label.debug();
+        label.setAlignment(Align.center);
+
+        return label;
+    }
+
+    private boolean shouldShowUseMessage() {
```
Comment (joshuaskelly) [inline Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/GameMessage.java:97]: A side effect of this method is it will new up a `LookAtDTO` object. Probably should cache that somewhere?

File: Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/GameMessage.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                super.act(delta);
+
+                if (!shouldShowUseMessage()) {
+                    setText("");
+                    setVisible(false);
+                    return;
+                }
+
+                setText(Game.instance.player.lookedAtItem.getLookAtInfo().getTitle());
+                setColor(Game.instance.player.lookedAtItem.getLookAtInfo().getTitleColor());
```
Comment (joshuaskelly) [inline Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/GameMessage.java:64]: +1 to newing up a DTO here as well. Basically anywhere you access `lookedAtItem.getLookAtInfo()` you are creating new objects.

Could the info potentially change? Either way, I think we can be a bit more careful with regards to GC. 🗑️

File: Dungeoneer/src/com/interrupt/dungeoneer/dto/LookAtDTO.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    private String title;
+    private String attributes;
+    private Color color;
+
+    public LookAtDTO(String title, String attributes, Color color) {
+        this.title = title;
```
Comment (joshuaskelly) [inline Dungeoneer/src/com/interrupt/dungeoneer/dto/LookAtDTO.java:11]: Since this is usually something like: `E: use` etc should we call this `prompt`?

File: Dungeoneer/src/com/interrupt/dungeoneer/dto/LookAtDTO.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        this.attributes = attributes;
```
Comment (joshuaskelly) [inline Dungeoneer/src/com/interrupt/dungeoneer/dto/LookAtDTO.java:12]: I wasn't super sure what attributes meant here. Maybe call this additional info?

File: Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/GameMessage.java (hunk 3 above)
Comment (joshuaskelly) [reply Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/GameMessage.java:97]: Also this feels like maybe it should be a method on the Game class?

File: Dungeoneer/src/com/interrupt/dungeoneer/dto/LookAtDTO.java (hunk 1 above)
Comment (PythooonUser) [reply Dungeoneer/src/com/interrupt/dungeoneer/dto/LookAtDTO.java]: Fixed.

File: Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/GameMessage.java (hunk 4 above)
Comment (PythooonUser) [reply Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/GameMessage.java]: Yes, info can change when e.g. a door opens or closes. But it is now cached in the player.

**Review by joshuaskelly — COMMENTED** [review]
Leaving a round of feedback. Works well and heading in the right direction. 👍

File: Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/DynamicFormatText.java [hunk 7]
```diff
@@ -13,6 +13,7 @@
 
 public class DynamicFormatText extends Element {
     public String pattern = "{0}";
+    public float frequency = 100f;
```
[the remark this answers is not in the document]
Comment (PythooonUser) [reply Dungeoneer/src/com/interrupt/dungeoneer/ui/layout/DynamicFormatText.java:16]: Try setting it to 10Hz and then look at objects. It lacks behind quite substantially.

File: Dungeoneer/src/com/interrupt/dungeoneer/entities/Entity.java [hunk 8]
```diff
@@ -17,12 +17,14 @@
...
 import com.interrupt.dungeoneer.gfx.drawables.DrawableSprite;
+import com.interrupt.dungeoneer.interfaces.LookAt;
+import com.interrupt.dungeoneer.interfaces.LookAtInfoModifier;
 import com.interrupt.dungeoneer.tiles.Tile;
 
 import java.util.Random;
 
 /** Base class for all entities in Delver Engine Levels. */
-public class Entity {
+public class Entity implements LookAt {
```
[the remark this answers is not in the document]
Comment (PythooonUser) [reply Dungeoneer/src/com/interrupt/dungeoneer/entities/Entity.java:27]: What is your suggestion here? Let specific sub-classes implement the interface instead?

File: Dungeoneer/src/com/interrupt/dungeoneer/game/Game.java [hunk 9]
```diff
@@ -72,8 +73,7 @@
 	public static float messageTimer = 0;
 	public static float messageScale = 1;
 	public static Array<String> message = new Array<String>();
-	public static Array<String> useMessage = new Array<String>();
-	public static Color useMessageColor = Color.WHITE;
+    public static Message message2 = new Message();
```
[the remark this answers is not in the document]
Comment (PythooonUser) [reply Dungeoneer/src/com/interrupt/dungeoneer/game/Game.java:76]: Yes, this is just a temporary naming right now.

File: Dungeoneer/src/com/interrupt/dungeoneer/entities/Entity.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (joshuaskelly) [reply Dungeoneer/src/com/interrupt/dungeoneer/entities/Entity.java:27]: Because the look at mechanic is gameplay specific, maybe some sort of helper class?

[the remark this answers is not in the document]
Comment (PythooonUser) [reply Dungeoneer/src/com/interrupt/dungeoneer/entities/Entity.java:27]: Mmh, not sure if I understand. What would a helper class do here?

I could only think of the following: We call a helper method whenever we want to get the look at info. This method takes in the object (not the interface anymore, could be tricky) and then has all the code to create that info in a giant switch statement for the actual object type/class.

So we would move all the ovject specific code out of the classes and move it into the helper.

Feels a bit clunky as well. The interface looked nice to be honest. Usually you always need a look at mechanic?!

What was your idea here?

[the remark this answers is not in the document]
Comment (joshuaskelly) [reply Dungeoneer/src/com/interrupt/dungeoneer/entities/Entity.java:27]: My thinking it is heavily tied towards Delver specific gameplay. If I wanted to make a top down twin stick shooter with Delver Engine (I do) this interface is meaningless.
